pinot-controller/src/main/java/com/linkedin/pinot/controller/api/restlet/resources/PinotSegmentRestletResource.java

```diff
@@ -230,19 +232,32 @@ protected Representation toggleSegmentState(String tableName, String segmentName
 
     JSONArray ret = new JSONArray();
     String tableNameWithType = null;
-    if ((tableType == null || TableType.OFFLINE.name().equalsIgnoreCase(tableType))
+    List<String> segmentsToToggle;
+
+    if (segmentName != null) {
+      segmentsToToggle = Collections.singletonList(segmentName);
+    } else {
+      segmentsToToggle = _pinotHelixResourceManager.getAllSegmentsForResource(tableName);
+    }
+    if ((tableType == null
+        && SegmentName.getSegmentType(segmentsToToggle.get(0)).equals(SegmentName.RealtimeSegmentType.UNSUPPORTED)
```

Comment:
I feel this method is doing a lot of work to handle tableName with or without type, type parameter and empty or non-empty segment list. This can also lead to confusing responses - if this is called with 
myTable_OFFLINE, realtime_segment_name, type==null then it will turn tableName to myTable_OFFLINE_REALTIME and throw error downstream.

Best to limit the contract and fix callers. Don't infer types from table names given customers can download segments and control uploading it with specific names. Keep the logic in this method simple. You can also consider providing wrapper methods to simplify logic to always provide tableType based on tableName...just to throw some options.

Also, please fix the controller api swagger doc to indicate if table name requires type and semantics